services/collaboration/pkg/internal/app/wopifilecontents.go

```diff
@@ -0,0 +1,169 @@
+	if err != nil {
+		app.Logger.Error().
+			Err(err).
+			Str("FileReference", wopiContext.FileReference.String()).
+			Str("RequestedLockID", r.Header.Get(HeaderWopiLock)).
+			Str("UploadLength", strconv.FormatInt(r.ContentLength, 10)).
+			Str("ViewMode", wopiContext.ViewMode.String()).
+			Str("Requester", wopiContext.User.GetId().String()).
+			Msg("PutFile: stat failed")
+		http.Error(w, http.StatusText(http.StatusInternalServerError), http.StatusInternalServerError)
+		return
+	}
+
+	if statRes.Status.Code != rpcv1beta1.Code_CODE_OK {
+		app.Logger.Error().
+			Str("FileReference", wopiContext.FileReference.String()).
+			Str("RequestedLockID", r.Header.Get(HeaderWopiLock)).
+			Str("UploadLength", strconv.FormatInt(r.ContentLength, 10)).
+			Str("ViewMode", wopiContext.ViewMode.String()).
+			Str("Requester", wopiContext.User.GetId().String()).
+			Str("StatusCode", statRes.Status.Code.String()).
+			Str("StatusMsg", statRes.Status.Message).
+			Msg("PutFile: stat failed with unexpected status")
+		http.Error(w, http.StatusText(http.StatusInternalServerError), http.StatusInternalServerError)
+		return
+	}
+
+	// If there is a lock and it mismatches, return 409
+	if statRes.Info.Lock != nil && statRes.Info.Lock.LockId != r.Header.Get(HeaderWopiLock) {
+		app.Logger.Error().
+			Str("FileReference", wopiContext.FileReference.String()).
+			Str("RequestedLockID", r.Header.Get(HeaderWopiLock)).
+			Str("UploadLength", strconv.FormatInt(r.ContentLength, 10)).
+			Str("ViewMode", wopiContext.ViewMode.String()).
+			Str("Requester", wopiContext.User.GetId().String()).
+			Str("LockID", statRes.Info.Lock.LockId).
+			Msg("PutFile: wrong lock")
+		// onlyoffice says it's required to send the current lockId, MS doesn't say anything
+		w.Header().Add(HeaderWopiLock, statRes.Info.Lock.LockId)
+		http.Error(w, http.StatusText(http.StatusConflict), http.StatusConflict)
+		return
+	}
+
+	// only unlocked uploads can go through if the target file is empty,
+	// otherwise the X-WOPI-Lock header is required even if there is no lock on the file
+	// This is part of the onlyoffice documentation (https://api.onlyoffice.com/editors/wopi/restapi/putfile)
+	// Wopivalidator fails some tests if we don't also check for the X-WOPI-Lock header.
+	if r.Header.Get(HeaderWopiLock) == "" && statRes.Info.Lock == nil && statRes.Info.Size > 0 {
+		app.Logger.Error().
+			Str("FileReference", wopiContext.FileReference.String()).
+			Str("RequestedLockID", r.Header.Get(HeaderWopiLock)).
+			Str("UploadLength", strconv.FormatInt(r.ContentLength, 10)).
+			Str("ViewMode", wopiContext.ViewMode.String()).
+			Str("Requester", wopiContext.User.GetId().String()).
+			Msg("PutFile: file must be locked first")
+		// onlyoffice says to send an empty string if the file is unlocked, MS doesn't say anything
+		w.Header().Add(HeaderWopiLock, "")
+		http.Error(w, http.StatusText(http.StatusConflict), http.StatusConflict)
+		return
+	}
+
+	// upload the file
+	err = helpers.UploadFile(
```

Comment:
this is a non atomic operation - unless I miss anything ...

there is a stat call above and here is the upload - there is a time frame in between these two where the state which was asserted based on the stat data could have changed.

no idea about the cs3 api capabilities - but this has to be wrapped in some kind of transaction or something ....

Reply:
I think this is covered by the WOPI locks, at least for the common cases, because locks are required. The only exception where you're allowed to upload without locks is when the target file is empty.

The only race condition that could happen is if the target file is empty and 2 unlocked uploads happen at the same time, but it shouldn't happen too often.

Reply:
"shouldn't happen too often" scares me 🤣

We should review all operations under this perspective - including implementations in reva which are executed.